### DISCUSSION TITLE
TODOs and Ideas

### DISCUSSION THREAD

**Discussion by salkinium** [modm-io/modm#708] [Ideas]
Some rough ideas about what needs to be done without having to open an individual issue for each one of them. Please feel free to comment or add more features.

## Library

- [ ] Run unittests in hardware on AVR and STM32 targets
- [ ] Fix AVR unittests
- [x] Automatic code formatting (via Clang-Format?)
- [x] Integrate [gdbgui](https://www.gdbgui.com)
- [ ] Move XPCC protocol out of modm into its own `xpcc-communication` repository.
- [ ] Refactor xpcc-generator into it's own proper python library
- [ ] Add [ufat](https://github.com/dlbeer/ufat) as alternative to `modm:fatfs`
- [ ] Integrate [`upainter` graphics algorithms](https://github.com/salkinium/upainter)
- [ ] Integrate [fixed point math](https://github.com/MikeLankamp/fpm)
- [ ] Let user choose which memory regions to use as heap
- [ ] Find a good queue library for multiple use-cases for Cortex-M/AVR
- [x] assign linkerscript sections via memory attributes
- [ ] call same function on a set of pins (ie. access-optimized `setAlternateFunction()`)
- [ ] Implement Semihost logging for boards without serial or SWO output
- [ ] Port fmtlib (aka. `std::format` in C++20)
- [ ] Port [defmt offline formatting library from Rust](https://github.com/knurling-rs/defmt)
- [ ] Project demo list as part of documentation
- [ ] Add release tags to CI containers to make it easier to pair them
- [ ] Add unit test via pyelftools to validate linkerscript assumptions, section placement/alignment/order/size.
- [ ] Better examples maybe generated
- [ ] Update documentation to describe how to add an external modular library to lbuild/Scons/CMake/Make
- [ ] Add documentation that points out best practices for drivers, board packages, HALs
- [x] Windows CI
- [ ] Use-case for Peripheral::connect() methods: e.g. combinations of SPI signals
  - only MO, no SCK: WS2812 accelerated bit-bang
  - MI and SCK: read-in 74HC165
  - MO and SCK: write-out 74HC595
  - Some SPI-slaves have Open-Drain pins -> optional PullUp
- [ ] Hook [hardware events directly via generated ISR multiplexors](https://github.com/modm-io/modm/issues/248#issuecomment-511884427) (attach functions to GPIOs directly)
- [x] Add USB stack by integrating one of these projects:
  - https://github.com/hathach/tinyusb
  - https://github.com/dmitrystu/libusb_stm32
  - https://github.com/IntergatedCircuits/USBDevice
  - https://github.com/kevinmehall/usb
- [ ] Integrate MRI: https://github.com/adamgreen/mri
- [ ] Better [default configuration structs for drivers](https://github.com/roboterclubaachen/xpcc/pull/86#issuecomment-186578136)
- [x] Generate a peripheral support matrix for all target families from lbuild data to replace the non-sensical "★" table in the main README.
- [x] Add stackful fiber implementation based on std::thread concurrency interface
- [x] Deprecate Protothreads/Resumables in favor of Fibers
- [ ] stringly-typed peripheral driver interface
- [x] Write CI test to check if all examples are run by the CI
- [ ] Add a porting guide, this issue is coming up a lot more now:
  - Can be based on the [comments for porting SAM* devices](https://github.com/modm-io/modm/issues/189#issuecomment-485135716)
- [ ] Add more heap allocator options for multi-pool heaps: https://github.com/MaJerle/lwmem
- [ ] Use [libusb1](https://pypi.org/project/libusb1/) in modm_tools to automatically choose programmer
- [ ] Integrate [pyOCD](https://github.com/pyocd/pyOCD) as an alternative to OpenOCD.
- [x] Add a `modm:ide:vscode` module to add [thorough support for Cortex-M devices](https://mcuoneclipse.com/2021/06/02/visual-studio-code-for-c-c-with-arm-cortex-m-part-7-freertos/)
- [x] Integrate J-Link support
- [ ] Check Doxypress `@ingroup` fits the module name that generated the file via lbuild buildlog

## Build

- [x] More flexible git info for any repository.

**Comment by asmfreak** [comment]
I saw some time ago discussion on code formatting.
I saw [this tool](https://zed0.co.uk/clang-format-configurator/) on C++ Weekly, enabling one to choose and save clang-format configuration.
I'm not sure, if we'll be able to make great use out of it, since a lot of modm code is Jinja templates, unparseable by clang.

**Comment by parzival2** [comment]
Any plans to add USB support?

**Comment by salkinium** [comment]
> Any plans to add USB support?

Yes, preferrably via third-party project integration though, since doing USB properly seems fairly complicated. I've bookmarked these projects to evaluate:

- https://github.com/hathach/tinyusb
- https://github.com/dmitrystu/libusb_stm32
- https://github.com/IntergatedCircuits/USBDevice
- https://github.com/kevinmehall/usb

**Comment by parzival2** [comment]
@salkinium Great. If I get around to gain a bit of knowledge into the project, may be I can create a pull-request using any of the above mentioned sources. I just discovered modm yesterday and its great.

**Comment by parzival2** [comment]
I saw that ROS is already available but having nanopb(https://github.com/nanopb/nanopb) will also be really great. It will be useful in writing binary files to SD-Card. The idea that I had in mind was to collect Accelerometer, Gyroscope data and play them back to simulate sensors like ROS bags. What do you guys think?

**Comment by salkinium** [comment]
> What do you guys think?

You don't *have* to integrate nanopb into modm directly (in fact I would *prefer* to keep the modm core repo not expanding in size too much). You could make it a separate lbuild repository (like https://github.com/modm-io/Invensense-eMD) or modify the *generated* SCons/CMake build system to also build nanopb. The goal is to not have to "wrap" every library with modm, as modm probably adds very little to those libraries.

> It will be useful in writing binary files to SD-Card

Note that our version of FatFS is very old, it ~should~ must be upgraded. In addition there isn't that much *explicit* support for SD-Cards or filesystems in general in modm. (sorry)

**Comment by salkinium** [comment]
Oh, nanopb comes with both CMake and SCons tools, so maybe it is warranted to create a proper lbuild repository and integrate not only the source but also the protocol builder into the modm build system. Probably not trivial though.

**Comment by W-M-D** [comment]
Implementing some SD support soon!

**Comment by rleh** [comment]
> Implementing some SD support soon!

Once I added a block device interface (and implementations for SPI flash, heap storage and mirroring (like RAID1) to modm: https://docs.modm.io/develop/api/stm32g474vet7/classmodm_1_1BlockDevice.html
The idea was to use this interface also for SD cards, on-chip flash and other storage devices.

Then file systems like FAT (or ext3) could be implemented via this interface and used for all storage.

Unfortunately this interface has not been used for the recently added Eeprom driver in #321.

**Comment by mikewolfram** [comment]
I wasn't aware of the BlockDevice, so I did the CAT24AAxx driver by simply modifying the existing EEPROM driver. I'll look into it and change it.

**Comment by mikewolfram** [comment]
Talking about upcoming implementations, I'm interested in DMA and UI. I've been looking in lvgl but wasn't able to get it up and running. Their memory management didn't work on my L431. And also it adds a lot of code, my test of having a simple button added >30k (with -Og setting). I'd rather use something else. The current UI in modm has small footprint and would work perfectly to my needs.

**Comment by rleh** [comment]
> DMA

You know there currently exists a very basic and hardly used [DMA API](https://docs.modm.io/develop/api/stm32f407zgt7/group__modm__platform__dma.html)?

@chris-durand and me started to work on DMA two years ago. On my [`feature/dmahal` branch](https://github.com/rleh/modm/commits/feature/dmahal) are a lot of *WIP*-commits. I'll try to rebase the branch soon.

DMA can be used for many different use cases. What is your use case?
Maybe we could discuss the use-cases and thus the interface of the DMA API(s) in a separate issue?

**Comment by salkinium** [comment]
>  I'm interested in DMA

<details>
<summary>Meme</summary>

![](https://user-images.githubusercontent.com/163066/77000627-fc59bf00-6958-11ea-9b7d-743c9095d54e.jpg)

</details>

> Maybe we could discuss the use-cases and thus the interface of the DMA API(s) in a separate issue?

Yes please. I've added a [bunch of DMA data already to the device files](https://github.com/modm-io/modm-devices/pull/1) for this purpose!

**Comment by rleh** [comment]
For DMA: see #358

**Comment by ghost** [comment]
Better sanity checks. After some quick testing i encountered type mismatches across units. Too many warnings is disabled after my taste, which can lead to undefined behaviour. Tools as clang-tidy and cppcheck can be used on the generated code.

Example of code smells:

_platform/fault/fault_storage.cpp:36:17: note: type 'const uint32_t' should match type 'const uint8_t'
   36 | const uint32_t *fault_storage_heap_start{nullptr};_

_platform/core/heap_table.cpp:22:27: warning: type 'struct table_pool_t' violates the C++ One Definition Rule [-Wodr]
   22 | } __attribute__((packed)) table_pool_t;_

_platform/fault/fault_storage.cpp:18:17: note: a field of same name but different type is defined in another translation unit
   18 |  uint32_t const traits;_

**Comment by salkinium** [comment]
I completely agree! This only works if it is automated in the CI though. Would you be interested in setting this up (or adding more warnings)?

**Comment by yandi-banyu** [comment]
Any plans on adding HAL_Ex function of STM32? For example, the HAL_UARTEx_ReceiveToIdle_DMA is quite useful when making a simple packet based serial communication. I usually implement a simple [COBS](https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing) encoded packet with CRC and send it away. When receiving I can just detect packet error when there is no 0x00 delimiter on the rx queue, then make sure the data is not corrupted by checking the CRC.
